File: Homework/Ngan_Luu/150-evaluate-reverse-polish-notation.py
https://leetcode.com/problems/evaluate-reverse-polish-notation/
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack=[]
        for token in tokens:
            if token not in ['+','-','*','/']:
                stack.append(int(token))
            else:
                if len(stack)<2:
                    return False
                else:
                    a=stack.pop()
                    b=stack.pop()
                    if token == '+':
                        stack.append(a+b)
                    elif token == '-':
                        stack.append(b-a)
                    elif token =='*':
                        stack.append(a*b)
                    else:
                        stack.append(int(b/a))
        return stack[0]

File: Homework/Son/W2/150-evaluate-reverse-polish-notation.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
 var evalRPN = function(tokens) {
    var math_it_up = {
        '+': function ([x, y]) { return x + y },
        '-': function ([x, y]) { return x - y },
        '*': function ([x, y]) { return x * y },
        '/': function ([x, y]) { 
            if (x/y > 0) return Math.floor(x / y) 
            else return Math.ceil(x/y)
        },
    };
    let numStack = []
    for (let token of tokens) {
        if (token !== "+" && token !== "-" && token !== "*" && token !== "/") {
            numStack.push(parseInt(token))} 
        else {
            let calculation = math_it_up[token](numStack.slice(-2))
            numStack.pop()
            numStack.pop()
            numStack.push(calculation)
        }
    }
    return numStack[0]
};

// https://leetcode.com/problems/evaluate-reverse-polish-notation/